### ISSUE TITLE
Implement training

### ISSUE THREAD

**Issue by andrewni2002** [andrewni420/GPU-PUSH#5]
Like with flax+optax, training should be centered around a TrainState object, whether that's a class or just a dictionary.

The training state should include 
- model parameters
- evaluation function
- optimization function

Each training step should take in
- training state
- inputs
- labels

and return
- updated training state
- (optional) training metrics, such as training loss, accuracy, etc.

Overall, the trainer should take in an unevaluated individual and return the individual with:
- Final parameter values
- Final train/validation loss/acc (or other fitness measure)
- Other training statistics (time taken, # params, memory, etc.)
